codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,412 +70,1049 @@
+ * note that only {@link Http2DataFrame}s are subject to HTTP/2 flow control. So it's perfectly legal (and expected)
+ * by a handler that aims to respect the channel's writability to e.g. write a {@link Http2DataFrame} even if the
+ * channel is marked unwritable.
  */
 @UnstableApi
-public final class Http2MultiplexCodec extends ChannelDuplexHandler {
+public class Http2MultiplexCodec extends Http2FrameCodec {
 
-    private static final InternalLogger logger = InternalLoggerFactory.getInstance(Http2MultiplexCodec.class);
+    private static final ChannelFutureListener CHILD_CHANNEL_REGISTRATION_LISTENER = new ChannelFutureListener() {
+        @Override
+        public void operationComplete(ChannelFuture future) throws Exception {
+            registerDone(future);
+        }
+    };
 
-    private final Http2StreamChannelBootstrap bootstrap;
+    private static final ChannelMetadata METADATA = new ChannelMetadata(false, 16);
+    private static final ClosedChannelException CLOSED_CHANNEL_EXCEPTION = ThrowableUtil.unknownStackTrace(
+            new ClosedChannelException(), DefaultHttp2StreamChannel.Http2ChannelUnsafe.class, "write(...)");
+    /**
+     * Number of bytes to consider non-payload messages, to determine when to stop reading. 9 is
+     * arbitrary, but also the minimum size of an HTTP/2 frame. Primarily is non-zero.
+     */
+    private static final int ARBITRARY_MESSAGE_SIZE = 9;
 
-    private final List<Http2StreamChannel> channelsToFireChildReadComplete = new ArrayList<Http2StreamChannel>();
-    private final boolean server;
-    private ChannelHandlerContext ctx;
-    private volatile Runnable flushTask;
+    /**
+     * Returns the flow-control size for DATA frames, and 0 for all other frames.
+     */
+    private static final class FlowControlledFrameSizeEstimator implements MessageSizeEstimator {
+
+        static final FlowControlledFrameSizeEstimator INSTANCE = new FlowControlledFrameSizeEstimator();
 
-    private final IntObjectMap<Http2StreamChannel> childChannels = new IntObjectHashMap<Http2StreamChannel>();
+        static final MessageSizeEstimator.Handle HANDLE_INSTANCE = new MessageSizeEstimator.Handle() {
+            @Override
+            public int size(Object msg) {
+                return msg instanceof Http2DataFrame ? ((Http2DataFrame) msg).flowControlledBytes() : 0;
+            }
+        };
+
+        @Override
+        public Handle newHandle() {
+            return HANDLE_INSTANCE;
+        }
+    }
+
+    private static final class Http2StreamChannelRecvByteBufAllocator extends DefaultMaxMessagesRecvByteBufAllocator {
+
+        static final Http2StreamChannelRecvByteBufAllocator INSTANCE = new Http2StreamChannelRecvByteBufAllocator();
+
+        @Override
+        public MaxMessageHandle newHandle() {
+            return new MaxMessageHandle() {
+                @Override
+                public int guess() {
+                    return 1024;
```

Comment:
This relates to https://github.com/netty/netty/pull/6896#discussion_r132092664. The entire RecvByteBufAllocator concept doesn't make much sense in the context of child channels. It was designed to help control the process of reading bytes from a socket, size the buffer when a read is done, and determine if we should keep reading. In this case the child channel is handed objects after this point. I think we should revisit this as a followup. Consider putting a TODO and creating a followup issue.